mne/preprocessing/_regress.py

```diff
@@ -53,35 +69,180 @@ def regress_artifact(inst, picks=None, picks_artifact='eog', betas=None,
+        -------
+        self : EOGRegression
+            The fitted ``EOGRegression`` object. The regression coefficients
+            are available as the ``.coef_`` and ``.intercep_`` attributes.
+
+        Notes
+        -----
+        If your data contains EEG channels, make sure to apply the desired
+        reference (see :func:`mne.set_eeg_reference`) before performing EOG
+        regression.
+        """
+        self._check_inst(inst)
+        picks = _picks_to_idx(inst.info, self._picks, none='data',
+                              exclude=self._exclude)
+        picks_artifact = _picks_to_idx(inst.info, self._picks_artifact)
+
+        # Calculate regression coefficients. Add a row of ones to also fit the
+        # intercept.
+        _check_preload(inst, 'artifact regression')
+        artifact_data = inst._data[..., picks_artifact, :]
+        ref_data = artifact_data - np.mean(artifact_data, -1, keepdims=True)
+        if ref_data.ndim == 3:
+            ref_data = ref_data.transpose(1, 0, 2)
+            ref_data = ref_data.reshape(len(picks_artifact), -1)
+        cov_ref = ref_data @ ref_data.T
+
+        # Process each channel separately to reduce memory load
+        coef = np.zeros((len(picks), len(picks_artifact)))
+        for pi, pick in enumerate(picks):
+            this_data = inst._data[..., pick, :]  # view
+            # Subtract mean over time from every trial/channel
             cov_data = this_data - np.mean(this_data, -1, keepdims=True)
             cov_data = cov_data.reshape(1, -1)
-            betas[pi] = np.linalg.solve(cov, np.dot(ref_data, cov_data.T)).T[0]
-        # subtract weighted (demeaned) eye channels from channel
-        this_data -= (betas[pi] @ ref_data).reshape(orig_shape)
-    return inst, betas
+            # Perform the linear regression
+            coef[pi] = np.linalg.solve(cov_ref, ref_data @ cov_data.T).T[0]
+
+        # Store relevant parameters in the object.
+        self.info = pick_info(inst.info, picks)
+        self.coef_ = coef
+        self._picks = [inst.ch_names[ch] for ch in picks]
+        self._picks_artifact = [inst.ch_names[ch] for ch in picks_artifact]
+        return self
+
+    @fill_doc
+    def apply(self, inst, taa=False, copy=True):
+        """Apply the regression coefficients to some data.
+
+        Parameters
+        ----------
+        inst : Raw | Epochs | Evoked
+            The data on which to apply the regression.
+        taa : bool
+            Whether to apply TAA correction as described in equation 9 of
+            :footcite:`CroftBarry2000`. Defaults to ``False``.
+        %(copy_df)s
+
+        Returns
+        -------
+        inst : Raw | Epochs | Evoked
+            A version of the data with the artifact channels regressed out.
+
+        Notes
+        -----
+        Only works after ``.fit()`` has been used.
+
+        References
+        ----------
+        .. footbibliography::
+        """
+        self._check_inst(inst)
```

Comment:
Would it? That would be bad indeed. The intent was that copy should come first.

Reply:
indeed it would. `_check_inst` has
```py
        if not inst.proj:
            inst.apply_proj()
```
and the copy doesn't happen until 7 lines below here.

Reply:
I just changed to behavior to not automatically apply projections. That is now driving me nuts. Whenever I want to do regression, I must first manually `raw.apply_proj()`. There are other places where MNE-Python automatically applies the projections (hence modifying the data in place), for example when making evokeds or computing inverses. That always has been one of the gotcha's of MNE-Python, but not doing it automatically is actually kind of annoying. @agramfort @larsoner I would appreciate your opinion on this as well.

Reply:
I would add a `proj=True` kwarg here, and make sure the conditional for `apply_proj` happens after the `copy`

Reply:
agreed. Except for inverse where I think it makes no sense to not apply them, we should be using a proj parameter that can be toggled to False if needed